### ISSUE TITLE
Human Handoff Status Update

### ISSUE THREAD

**Issue by vishnu2017** [Microsoft/BotBuilder-Samples#1248]
Do we have any plans of rolling out the human hand off feature as a sample for node and C# - **GA**

**Comment by v-kydela** [comment]
Do these work?

C# - https://github.com/tompaana/intermediator-bot-sample
Node - https://github.com/GeekTrainer/botframework-v4-handoff

**Comment by 01** [comment]
Yes the botframework-v4-handoff works

**Comment by v-kydela** [comment]
@vishnu2017 - Does that resolve your issue?

**Comment by v-kydela** [comment]
@vishnu2017 - Are you still working on this?

**Comment by v-kydela** [comment]
Closing due to inactivity

**Comment by EugeneFoo** [comment]
Can please re-open this issue?

As bot framework V4 keep changing and seem that repo doesn't have any update anymore.

Node - https://github.com/GeekTrainer/botframework-v4-handoff

Or does it provide official sample for this bot-handoff?

**Comment by v-kydela** [comment]
@EugeneFoo - You can go ahead and make a feature request in this repo. Make sure you provide the information in the [feature request template](https://github.com/Microsoft/BotBuilder-Samples/issues/new?template=-samples-feature-request.md).

**Comment by EugeneFoo** [comment]
> @EugeneFoo - You can go ahead and make a feature request in this repo. Make sure you provide the information in the [feature request template](https://github.com/Microsoft/BotBuilder-Samples/issues/new?template=-samples-feature-request.md).

Thanks @v-kydela

**Comment by JohnGalt1717** [comment]
The problem with the current samples (and lack of ability to easily do this now without major additional libraries and hacking) is the following:

1. The sample is overly complex so even if you do figure it out it will have taken you hours to do something that should be automatic.
2. It requires active subscription to hand-off with no way of having it hand off to a community of online users.
3. The same as issues (which I've documented in that repro) with #2 even if you can make it work.

There should be multiple samples and this should be easily added to Bot Framework in general for:

1. Hand-off to a group in Teams based on online status.
2. The ability to do custom logic to filter the list of people online. (i.e. experts based on the question asked)
3. Ability to track users that are currently online and do #2 against them. (i.e. community expert hand-off)

In general, this is the natural progression of Q&A Maker and needs to be drastically flushed out to work better and easier in Bot Framework because it is a vital function for a truly useful bot.  This can be a major differentiator between Azure Bot Framework and the equivalents from competitors. Or it can be the major lacking feature that makes companies like ours choose their's.

As it stands right now I've had to create our own bot system using SignalR (which is what Bot Framework should be using BTW) to be able to get the functionality that we, and I strongly suspect anyone doing something serious would need.  Using SignalR I can easily tell the people that are online or have dropped off (because of WebSocket disconnects on browser close) which lets me do true welcome messages instead of having to wait for the user to post a message.  Using SignalR I can easily push to other people in batches to request assistance and keep track of where they're at.  And of course, instead of using POST I'm using a real, realtime communication system.

The downside is that it doesn't work easily with Teams, Slack etc. but we're now, because hand-off is so poor in Bot Framework considering adding it to ours instead of waiting for Bot Framework to get to where it needs to be.

**Comment by 01** [comment]
@JohnGalt1717 What specific issues are you having? 

Are you frustrated at Microsoft not having prepackaged solutions open source immediately available for your exact use case or that the bot framework isnt able to do something you are trying to accomplish? 

Everything you listed is completely doable, the only one is proactively messaging without a user reference (this is frustrating and completely agree but for your use case of specific list of people this is easily accomplished bit hard if you have an organization of 10,000).

So ill try to help you since I had to bang my head against the problem to develop a solution very similiar in node js to what you are trying to accomplish

> The sample is overly complex so even if you do figure it out it will have taken you hours to do something that should be automatic.
**Cant help you here this a matter of opinion**
> It requires active subscription to hand-off with no way of having it hand off to a community of online users.
**Create a team for those you want want to be in the "community of online users" and at the bot, you now how a reference to the whole team roster including ids to proactively message**
> The same as issues (which I've documented in that repro) with #2 even if you can make it work.
**Cant actually see what issue is it just a copy of this one **
> There should be multiple samples and this should be easily added to Bot Framework in general for:
> **Multiple samples of what?**
> Hand-off to a group in Teams based on online status.
**So i dont have straight forward solution to this, my use case has agents logging in to bot to distinguish just being online vs being available to take requests, but for everything but online status put the "experts" team and you have access to the teams roster will have all ChannelAccount information needed to message them, im sure the teams API probably has a way to get online status, otherwise you could message all relevant experts and delete the message after some amount of time without response**
> The ability to do custom logic to filter the list of people online. (i.e. experts based on the question asked) 
>**What is preventing you from implementing custom logic, you will need to store the data about experts somewhere, you can use extensions to add custom data to user object using the graph api https://docs.microsoft.com/en-us/graph/extensibility-overview**
> Ability to track users that are currently online and do #2 against them. (i.e. community expert hand-off)
**For a work around have expert "login" to the bot and just store it in user or conversation state**

**Comment by JohnGalt1717** [comment]
@01 

Thanks for the replies. 

There's no issue with Q&A Maker persay but there is basically no use case where a company is going to stop there. They'll absolutely have escalation and it should be trivial to implement unlike the complex mess it is.

My comments about hand-off being brutally complex are based on just how involved the sample code is (and the documentation doesn't even link to it properly btw).

I should be able to do this:

Lookup userId from database of preferred escalations filtered by those on line and pick the best that isn't in another conversation.  

This should be a simple list of online users. The teams/Slack/etc integration should automatically register against the bot service list of online users based on criteria (i.e. available and in a group).

I should be able to easily register other online users that are in chats in other places that updates that list.

Users should all be easily identifiable having had their credentials decoded using ASP.NET Core identity (i.e. pulled their principle and claims from an OpenIdConnect token passed in the Authorization header as a bearer token). (I shouldn't have to write ANY code at all for this. It should just automatically happen as part of the Bot Framework and identity.)

The list should establish a heartbeat so that users no longer online are automatically culled from the list of users. For web this should be done with a websocket, for directline pick your poison and for teams etc. the user list should be subscribing to service's events for availability etc.  Again, I shouldn't have to do anything to make this happen, it should just work and raise events.  The issue right now is that I can register when users want to take questions, but it's basically impossible to remove them when they close the browser etc. so I have to fail, and I can't reach out unless there is actually a session.

Adding a user to a conversation should be no more complex than Recipeints.Add(<userId>) and I should be able to filter the conversation to block commands from being sent to the other side etc. using a simple event.

This is why we ultimately gave up on Azure Bot Framework the last time and did our own copying many of the principles but using SignalR. It solved all of the availability issues, and the identity issues automatically and made adding someone to a conversation easy and we got an event when they disconnect too.  

And since this is very much an RTC style communication I can't for the life of me figure out why Azure Bot Framework doesn't use SignalR underneath.

But even if it doesn't, MS needs to make this so that there is a central list of users and you can easily send a registration message with credentials that automatically registers them properly, and make adding a human to a session no more complex than the above.  Everything else is noise that blocks adoption for more than the playtoy implementation. (and linking in IVR through Team Phone System is the next logical component along those lines too)

**Comment by 01** [comment]
@JohnGalt1717 

I mean I dont know how to tell you that you are describing a pretty specific use case. You can accomplish all this with the bot framework, but not sure what basis there is to think these these specific use cases should be top priority. 

> There's no issue with Q&A Maker persay but there is basically no use case where a company is going to stop there.

Well ya and there are teams at Microsoft continually expanding upon  virtually everything in the framework. Essentially it seems you feel that they should have every possible use case and functionality already developed and pre packaged today? You can see they are clearly expanding upon QnA maker. But you know as software engineer we do occasionally need to create a solution not just use one.
https://github.com/microsoft/BotBuilder-Samples/tree/master/experimental

> Lookup userId from database of preferred escalations filtered by those on line and pick the best that isn't in another conversation.

You can do do this? There are a number of ways to do this but in all of them you are going to have to do session tracking.

> This should be a simple list of online users. The teams/Slack/etc integration should automatically register against the bot service list of online users based on criteria (i.e. available and in a group).

This will never happen. Some of us are waiting on a way to automatically register teams users within our tenant that we have admin rights on. But For good reason the bot framework will never allow for proactively messaging any user what so ever

> Users should all be easily identifiable having had their credentials decoded using ASP.NET Core identity (i.e. pulled their principle and claims from an OpenIdConnect token passed in the Authorization header as a bearer token). (I shouldn't have to write ANY code at all for this. It should just automatically happen as part of the Bot Framework and identity.)

This is completely a matter of opinion and makes huge assumptions about how all their other customers use their product. Do I think that is relatively common use case? Sure.... do I think that it so vital that it must be implemented immediately as a part of the botbuilder core library asap, no there zero basis for that claim

> The list should establish a heartbeat so that users no longer online are automatically culled from the list of users. For web this should be done with a websocket, for directline pick your poison and for teams etc. the user list should be subscribing to service's events for availability etc. Again, I shouldn't have to do anything to make this happen, it should just work and raise events. 

I think this might be where the disconnect is happening for you. We are talking about dozen(s) of different chat protocol. The bot service is not designed with any of them specifically in mind,  but Microsoft has the bot connector so you dont have to be designing wrapper for every chat protocol simply to send a message. The more specific chat protocol information you want, teams, users, online status, slack specific, teams specific etc the more complicated that becomes. The botbuilder-teams team (OfficeDev) has provided libraries that expand upon that for teams specific stuff, but this is simply not the purpose of the bot service or framework itself. They provide a websocket for directline access to the bot service but its a bit of a stretch to expect then to provide that for other chat protocols. You have to understand even within Microsoft the OfficeDev team (MS team) is completely different then the people developing the bot framework stuff.  Dont get me wrong I have plenty of complaints about Teams right now but not going to yell at the bot sample team about it.

> This is why we ultimately gave up on Azure Bot Framework the last time and did our own copying many of the principles but using SignalR. It solved all of the availability issues, and the identity issues automatically and made adding someone to a conversation easy and we got an event when they disconnect too.

Fair enough Microsoft's Bot Service and associate framework isnt the holy grail of chat bot creation. Im sure that there a plenty use cases and scenarios were completely custom solutions using other technologies is the better option. Their product isnt for those. To provide what you want means that they arent working on something else, and you really dont have much of a basis for a claim that more customers want your specific functionality than other things that are being worked on.

> But even if it doesn't, MS needs to make this so that there is a central list of users and you can easily send a registration message with credentials that automatically registers them properly, and make adding a human to a session no more complex than the above.

As I mentioned you are never going to be able to proactively message any user you want. There is too much potential for abuse. As for adding a human to a sessions, it pretty easy and and can be done with a few line of code given you have the conversation references. You are conflating the handoff with all the internal logic and session tracking that is use case specific. Registration can be made easier with a simple middleware so you are just storing references on every message.

**Comment by Tushar301** [comment]
Hi @v-kydela, that tompanna solution helps in establishing 1:1 customer-agent communication, But how we'll established 1 to n (one agent handles multiple customers at a time)?

**Comment by 01** [comment]
@Tushar301 utilize the back channel to send identifying information for the agent and which customer/session the chat is for

**Comment by Tushar301** [comment]
Thanks for your reply @Andrew. Can you please tell me what changes I changes that I have to do in tompanna solution ?

**Comment by arman-g** [comment]
> Do these work?
> 
> C# - https://github.com/tompaana/intermediator-bot-sample
> Node - https://github.com/GeekTrainer/botframework-v4-handoff

The C# example is outdated. Would be nice to have it updated to .Net 5 and bot Framework v4.11 least. It seems the current sample might have custom implementation that is the integral part of the bot framework latest version. Some documentation would be great as to how exactly we can have a bot hand-off to a live agent in MS Teams. All I find is really concepts without any reasonable examples for a starter.

**Comment by 01** [comment]
@arman-g Im not familiar with C# as im using node.js but this seems to be the most up to date version on what microsoft recomends for handoff https://github.com/microsoft/BotBuilder-Samples/tree/main/experimental/handoff-library

**Comment by arman-g** [comment]
Hi @01, thanks for your response. Yes I saw that example but that is not what I am looking for. That example uses handoff integration model "Bot as an agent". What I am looking is "Bot as a proxy" since our clients use Web Chat channel talking to the bot and escalating to a live agent if there is a need using MS Teams as the agent hub.
